Attempt to Fix Scheduler

Nyefari opened the pull request:
- Add configuration item for preferred start time of day

- Adds check for if a challenge ended within 24 hours when the bot is started (If there is no current challenge). If there was, the bot waits for the next start time and starts the challenge then. If there wasn't , starts a new challenge.

- Changes the End time to always be challenge duration hours after the preferred start time of the day the challenge was started.

Nyefari committed:
fixed times I used single quote rather than backtick

Nyefari committed:
no message

Nyefari committed:
no message

Nyefari committed:
Merge branch 'master' of https://github.com/martindevans/Yolol-Ladder

# Conflicts:
#	YololCompetition/Services/Messages/DBMessages.cs

Nyefari committed:
Merge pull request #4 from martindevans/master

Hopefully updating my fork from Martin's Fork.

Nyefari committed:
Attempt to Fix Scheduler

- Add configuration item for preferred start time of day

- Adds check for if a challenge ended within 24 hours when the bot is started (If there is no current challenge). If there was, the bot waits for the next start time and starts the challenge then. If there wasn't , starts a new challenge.

- Changes the End time to always be challenge duration hours after the preferred start time of the day the challenge was started.

martindevans commented on YololCompetition/Services/Challenge/DbChallenges.cs:
```diff
@@ -90,7 +90,8 @@ DbCommand PrepareQuery(IDatabase db)
                 return null;
 
-            var endTime = (DateTime.UtcNow + TimeSpan.FromHours(_config.ChallengeDurationHours)).UnixTimestamp();
+            // End time of the challenge should be 3 days after the start time of today to account for if Referee was down on the day it was supposed to start the challenge.
```
> should be 3 days

This comment doesn't seem right?

Nyefari replied:
This is intentional. Same reason why if it has been over 24 hours when referee is restarted it will start a challenge regardless of what time it is.

martindevans commented on YololCompetition/Services/Challenge/IChallenges.cs:
```diff
@@ -25,6 +25,8 @@ public interface IChallenges
 
         Task<int> Delete(ulong challengeId);
+        
+        Task<Challenge?> GetChallengesByEndTime(ulong EndUnixTime);
```
parameters should start with a lower case letter: `endUnixTime`

martindevans replied:
Also, since this only gets a single challenge, it should be singular `GetChallengeByEndTime`

martindevans replied:
Actually looking more closely at the implementation, the name is misleading. See comments on impl.

Nyefari replied:
Fixed. Intentionally gets more than 1 to possibly have more uses for the command (get all challenges run in the last year, etc.)

martindevans commented on YololCompetition/Services/Challenge/DbChallenges.cs:
```diff
@@ -167,5 +168,13 @@ public async Task<int> Delete(ulong challengeId)
+            await using var cmd = _database.CreateCommand();
+            cmd.CommandText = "SELECT * FROM Challenges WHERE EndUnixTime >= @EndTime";
+            cmd.parameters.Add(new SqliteParameter("@EndTime", DbType.UInt64) { Value = EndUnixTime });
+            return await cmd.ExecuteNonQueryAsync();
```
There are a number of issues here:
 - You should probably use the `SqlAsyncResult` (see `GetPending` for example).
 - The name implies to be it gets a challenge by an exact end time, instead this will fetch all challenges ending after a given time.
 - They won't be in order!
 - The final call is `Execute**_Non_**QueryAsync`, so I'm not sure what this will do or if it actually compiles.

Nyefari replied:
Good shouts. Fixed.

martindevans commented on YololCompetition/Services/Schedule/InMemoryScheduler.cs:
```diff
@@ -56,24 +58,50 @@ public async Task Start()
-                    var next = await _challenges.StartNext();
-                    if (next == null)
+                    var currentTime = DateTime.UtcNow;
+                    var startTime = (currentTime.Date).AddMinutes(_config.ChallengeStartTime);  //Gets the time today that the challenge would be starting if it were starting today
```
As a matter of style I prefer `thing + TimeSpan.FromMinutes(N)` over `thing.AddMinutes(N)`. This is because it's not clear if `AddMinutes` mutates `thing` or not. Not critical though, just a style choice.

Nyefari replied:
This is why I included that () around currentTime.Date, to imply that it does not mutate the original thing. I understand your point though, and agree in principle. I'll change this and the other occurrences of similar.

martindevans commented on YololCompetition/Services/Schedule/InMemoryScheduler.cs:
```diff
@@ -56,24 +58,50 @@ public async Task Start()
-                        continue;
+
+                        // Start the next challenge, if there isn't one wait a while
+                        var next = await _challenges.StartNext();
```
If 24 hours have elapsed since the last challenge ended this starts a new one immediately, instead of waiting for the next start time.

Nyefari replied:
This is intentional. I was thinking that challenge uptime is more important than that the challenge start at the exact right time.

Nyefari replied:
This discussion is the one I'm waiting on before committing the other changes. 

The changes to the challenge start next is directly related to this decision. The start time used in the start next is based on UTC, and UTC wise we are likely to start in the final 7 hours of the day (sometime between 1700 UTC and 2000 UTC imo). Since the start time + duration is the start time of "today" + duration, if the current time when starting the challenge is after the start time of day, then the duration of the challenge will be slightly less. If it is earlier in the day than the start time then the duration will be longer. This likely gives us a range of 65 hours to 92 hours, with the bot automatically fixing the schedule to be based around the configured start time and duration for future challenges.

martindevans replied:
I think it's reasonable to leave it as it is. If one starts early (due to the very rare event that Referee was down for >24 hours) the one _after_ that will start on time and fix the schedule so overall it's really not worth worrying about.

Nyefari committed:
Changes requested during PR Review

Changes requested during PR Review
- using TimeSpan.From for addition/subtraction involving datetime.

- update the GetChallengesByEndDate to actually return the type it actually returns.

Nyefari committed:
Update InMemoryScheduler.cs

Final fix for thing I never saw the comment in the review for.
